cpp/src/lists/drop_list_duplicates.cu

```diff
@@ -165,6 +203,193 @@ std::unique_ptr<column> generate_entry_list_offsets(size_type num_entries,
+                          offset_type* output_begin,
+                          null_equality nulls_equal,
+                          nan_equality nans_equal,
+                          bool has_nulls,
+                          rmm::cuda_stream_view stream) const noexcept
+  {
+    if (nans_equal == nan_equality::ALL_EQUAL) {
+      list_entry_comparator<Type, true> const comp{list_offsets, d_view, nulls_equal, has_nulls};
+      return thrust::unique_copy(rmm::exec_policy(stream),
+                                 thrust::make_counting_iterator(0),
+                                 thrust::make_counting_iterator(num_entries),
+                                 output_begin,
+                                 comp);
+    } else {
+      list_entry_comparator<Type, false> const comp{list_offsets, d_view, nulls_equal, has_nulls};
+      return thrust::unique_copy(rmm::exec_policy(stream),
+                                 thrust::make_counting_iterator(0),
+                                 thrust::make_counting_iterator(num_entries),
+                                 output_begin,
+                                 comp);
+    }
+  }
+};
+
+/**
+ * @brief Copy list entries and entry list offsets ignoring duplicates
+ *
+ * Given an array of all entries flattened from a list column and an array that maps each entry to
+ * the offset of the list containing that entry, those entries and list offsets are copied into
+ * new arrays such that the duplicated entries within each list will be ignored.
+ *
+ * @param all_lists_entries    The input array containing all list entries
+ * @param entries_list_offsets A map from list entries to their corresponding list offsets
+ * @param nulls_equal          Flag to specify whether null entries should be considered equal
+ * @param nans_equal           Flag to specify whether NaN entries should be considered as equal
+ * value (only applicable for floating point data column)
+ * @param stream               CUDA stream used for device memory operations and kernel launches
+ * @param mr                   Device resource used to allocate memory
+ *
+ * @return A pair of columns, the first one contains unique list entries and the second one
+ * contains their corresponding list offsets
+ */
+std::vector<std::unique_ptr<column>> get_unique_entries_and_list_offsets(
+  column_view const& all_lists_entries,
+  column_view const& entries_list_offsets,
+  null_equality nulls_equal,
+  nan_equality nans_equal,
+  rmm::cuda_stream_view stream,
+  rmm::mr::device_memory_resource* mr)
+{
+  auto const num_entries    = all_lists_entries.size();
+  auto const d_view_entries = column_device_view::create(all_lists_entries, stream);
+
+  // Allocate memory to store the indices of the unique entries
+  auto const unique_indices = cudf::make_numeric_column(
+    entries_list_offsets.type(), num_entries, mask_state::UNALLOCATED, stream);
+  auto const unique_indices_begin = unique_indices->mutable_view().begin<offset_type>();
```

Comment:
You could just use a `device_uvector` here instead of having to make a column. Then in the gather call you can just use the iterator based `detail::gather` API.